source/particle/interpolator/nearest_neighbor.cc

```diff
@@ -0,0 +1,163 @@
+
+        std::vector<std::vector<double> > point_properties;
+
+        for (unsigned int pos_idx=0; pos_idx < positions.size(); ++pos_idx)
+          {
+            double minimum_distance = std::numeric_limits<double>::max();
+            if (n_particles > 0)
+              {
+                Particle<dim> nearest_neighbor;
+                for (typename std::multimap<types::LevelInd, Particle<dim> >::const_iterator particle = particle_range.first;
+                     particle != particle_range.second; ++particle)
+                  {
+                    const double dist = positions[pos_idx].distance(particle->second.get_location());
+                    if (dist < minimum_distance)
+                      {
+                        minimum_distance = dist;
+                        nearest_neighbor = particle->second;
+                      }
+                  }
+                std::vector<double> properties (n_particle_properties,0.0);
+                const dealii::ArrayView<const double> neighbor_props = nearest_neighbor.get_properties();
+                for (unsigned int i = 0; i < n_particle_properties; ++i)
+                  if (selected_properties[i])
+                    properties[i] = neighbor_props[i];
+                point_properties.push_back(properties);
+              }
+            else
+              {
+                std::vector<typename parallel::distributed::Triangulation<dim>::active_cell_iterator> neighbors;
+                GridTools::get_active_neighbors<parallel::distributed::Triangulation<dim> >(found_cell,neighbors);
+
+                int nearest_neighbor_cell = -std::numeric_limits<int>::max();
+                for (unsigned int i=0; i<neighbors.size(); ++i)
+                  {
+                    // Only recursively call this function if the neighbor cell contains
+                    // particles (else we end up in an endless recursion)
+                    if ((neighbors[i]->is_locally_owned())
+                        && (particles.count(std::make_pair(neighbors[i]->level(),neighbors[i]->index())) == 0))
+                      continue;
+                    else if ((!neighbors[i]->is_locally_owned())
+                             && (particle_postprocessor->get_particle_world().get_ghost_particles().count(
+                                   std::make_pair(neighbors[i]->level(),neighbors[i]->index())) == 0))
+                      continue;
+
+                    const double dist = positions[pos_idx].distance(neighbors[i]->center());
+                    if (dist < minimum_distance)
+                      {
+                        minimum_distance = dist;
+                        nearest_neighbor_cell = i;
+                      }
+                  }
+                std::vector<double> properties (n_particle_properties,0.0);
+                const std::vector<double> neighbor_props = properties_at_points(particles,
```

Comment:
I think you can just assign the return value to `point_properties`. It is the job of the recursively called function to select the appropriate components already.